src/hash.cpp

```diff
@@ -0,0 +1,156 @@
+    switch (digestType) {
+        case DigestTypes::SHA256:
+            return lib::OpenSSLLib::SSL_EVP_sha256();
+        case DigestTypes::SHA512:
+            return lib::OpenSSLLib::SSL_EVP_sha512();
+        default:
+            throw std::runtime_error("Unknown digest type");
+    }
+}
+
+static EVP_MD_CTX _initCtx (DigestTypes digestType) {
+    const EVP_MD* digestFn = _getDigestFn(digestType);
+    EVP_MD_CTX digestCtx;
+    EVP_MD_CTX_init(&digestCtx);
+    EVP_DigestInit_ex(&digestCtx, digestFn, NULL);
+    return digestCtx;
+}
+
+static void _finalizeDigest (EVP_MD_CTX& digestCtx, uint8_t* outDigest, unsigned int &outLength) {
+    EVP_DigestFinal_ex(&digestCtx, outDigest, &outLength);
+    EVP_MD_CTX_cleanup(&digestCtx);
+}
+
+static void _hash(DigestTypes digestType, const uint8_t* message, const int &length, uint8_t* outDigest, unsigned int &outLength) {
+    EVP_MD_CTX digestCtx = _initCtx(digestType);
+    EVP_DigestUpdate(&digestCtx, message, length);
+    _finalizeDigest(digestCtx, outDigest, outLength);
+}
+
+static void _hash(DigestTypes digestType, const std::string &message, std::string &outDigest) {
+    uint8_t value[Hash::lengthBytes[digestType]];
+    unsigned int length;
+    _hash(digestType, reinterpret_cast<const uint8_t*>(message.c_str()), message.length(), value, length);
+    outDigest = utility::toHex(value, length);
+}
+
+static void _hash(DigestTypes digestType, const std::vector<uint8_t> &message, std::vector<uint8_t> &outDigest) {
+    uint8_t value[Hash::lengthBytes[digestType]];
+    unsigned int length;    
+    _hash(digestType, &message[0], message.size(), value, length);
+    outDigest.insert(outDigest.begin(), value, value + length);
+    outDigest.resize(length);
+}
+
+static const std::string SHA256 = "sha256";
+static const std::string SHA512 = "sha512";
+
+void sha256(const uint8_t* message, const int &length, uint8_t* outDigest, unsigned int &outLength) {
+    _hash(DigestTypes::SHA256, message, length, outDigest, outLength);
+}
+
+void sha256(const std::string &message, std::string &outDigest) {
+    _hash(DigestTypes::SHA256, message, outDigest);
+}
+
+void sha256(const std::vector<uint8_t> &message, std::vector<uint8_t> &outDigest) {
+    _hash(DigestTypes::SHA256, message, outDigest);
+}
+
+void sha512(const uint8_t* message, const int &length, uint8_t* outDigest, unsigned int &outLength) {
+    _hash(DigestTypes::SHA512, message, length, outDigest, outLength);
+}
+
+void sha512(const std::string &message, std::string &outDigest) {
+    _hash(DigestTypes::SHA512, message, outDigest);
+}
+
+void sha512(const std::vector<uint8_t> &message, std::vector<uint8_t> &outDigest) {
+    _hash(DigestTypes::SHA512, message, outDigest);
+}
+
+std::map<DigestTypes, int> Hash::lengthBytes = {
+    { DigestTypes::SHA256, 256 / 8 },
+    { DigestTypes::SHA512, 512 / 8 }
+};
+
+Hash::Hash(const DigestTypes _digestType) : _digestType(_digestType) {
+    _digestCtx = _initCtx(_digestType);
```

Comment:
We don't have RAII pattern on the context. Consequently, there is a leak if the hash is (for whatever reason) never finalized. Use RAII here to avoid this.